moabb/paradigms/motor_imagery.py

```diff
@@ -82,43 +85,109 @@ def scoring(self):
-    def datasets(self):
-        return utils.dataset_search(paradigm='imagery',
-                                    events=['right_hand', 'left_hand'],
-                                    has_all_events=True)
+            # ensure events are desired:
+            if len(events) > 0:
+                keep_events = {key: val for key, val in event_dict.items() if
+                               val in np.unique(events[:, 2])}
+                if len(keep_events) > 0:
+                    # copy before filtering, so we let raws intact.
+                    for band_ind, (bp_low, bp_high) in enumerate(self.fbands):
+                        raw_f = raw.copy().filter(bp_low, bp_high, method='iir',
+                                                picks=picks, verbose=False)
+                        epochs = mne.Epochs(raw_f, events, event_id=keep_events,
+                                            tmin=time[0], tmax=time[1], proj=False,
+                                            baseline=None, preload=True,
+                                            verbose=False, picks=picks)
+                        out_dict[band_ind].append(epochs)
+
+        return out_dict
+    
+def ImageryNClassFactory(parent):
+
+    class ImageryNClass(parent):
+        """Imagery for multi class classification
+
+        Returns n-class imagery results, visualization agnostic but forces all
+        datasets to have exactly n classes. Uses 'accuracy' as metric
+
+        """
+
+        def __init__(self, n_classes, **kwargs):
+            self.n_classes = n_classes
+            super().__init__(**kwargs)
+
+        def verify(self, d):
+            log.warning(
+                'Assumes events have already been selected per dataset')
+            super().verify(d)
+            assert len(d.selected_events) == self.n_classes
+
+        @property
+        def datasets(self):
+            return utils.dataset_search(paradigm='imagery',
+                                        total_classes=self.n_classes,
+                                        has_all_events=True)
+
+    return ImageryNClass
+
+
+def LeftRightImageryFactory(parent):
+    class LeftRightImagery(parent):
+        """Motor Imagery for left hand/right hand classification
+
+        Metric is 'roc_auc'
+
+        """
+
+        def verify(self, d):
+            events = ['left_hand', 'right_hand']
+            super().verify(d)
+            assert set(events) <= set(d.event_id.keys())
+            d.selected_events = dict(
+                zip(events, [d.event_id[s] for s in events]))
+
+        @property
+        def scoring(self):
+            return 'roc_auc'
+
+        @property
+        def datasets(self):
+            return utils.dataset_search(paradigm='imagery',
+                                        events=['right_hand', 'left_hand'],
+                                        has_all_events=True)
+    return LeftRightImagery
+
+
+globals()['LeftRightImagerySinglePass'] = LeftRightImageryFactory(
```

Comment:
Factories to generate all valid classes as both SinglePass (one bandpass) and MultiPass